src/pymor/algorithms/rand_la.py

```diff
@@ -2,20 +2,183 @@
 # License: BSD 2-Clause License (https://opensource.org/licenses/BSD-2-Clause)
 
+from numbers import Number
 import numpy as np
 import scipy as sp
 from scipy.sparse.linalg import eigsh, LinearOperator
 from scipy.linalg import lu_factor, lu_solve
 from scipy.special import erfinv
 
 from pymor.algorithms.gram_schmidt import gram_schmidt
+from pymor.core.cache import CacheableObject, cached
 from pymor.core.defaults import defaults
-from pymor.operators.interface import Operator
 from pymor.core.logger import getLogger
+from pymor.operators.interface import Operator
 from pymor.operators.constructions import InverseOperator, IdentityOperator
+from pymor.tools.deprecated import Deprecated
+from pymor.tools.random import get_seed_seq, new_rng
+
+
+class RandomizedRangeFinder(CacheableObject):
+    def __init__(self, A, subspace_iterations=0, range_product=None, source_product=None, lambda_min=None,
+                 complex=False):
+        assert isinstance(A, Operator)
+        assert 0 <= subspace_iterations and isinstance(subspace_iterations, int)
+        if range_product is None:
+            range_product = IdentityOperator(A.range)
+        if source_product is None:
+            source_product = IdentityOperator(A.source)
+        assert isinstance(range_product, Operator)
+        assert isinstance(source_product, Operator)
+        assert lambda_min is None or isinstance(lambda_min, Number)
+        assert isinstance(complex, bool)
+
+        self.__auto_init(locals())
+        self._l = 0
+        self._Q = [self.A.range.empty()]
+        for _ in range(subspace_iterations):
+            self._Q.append(self.A.source.empty())
+            self._Q.append(self.A.range.empty())
+        self._Q = tuple(self._Q)
+        self.testvecs = self.A.source.empty()
+        self._basis_rng_real = new_rng(get_seed_seq().spawn(1)[0])
+        self._test_rng_real = new_rng(get_seed_seq().spawn(1)[0])
+        if complex:
+            self._basis_rng_imag = new_rng(get_seed_seq().spawn(1)[0])
+            self._test_rng_imag = new_rng(get_seed_seq().spawn(1)[0])
+
+    @cached
+    def _lambda_min(self):
+        if isinstance(self.source_product, IdentityOperator):
+            return 1
+        elif self.lambda_min is None:
+            with self.logger.block('Estimating minimum singular value of source_product...'):
+                def mv(v):
+                    return self.source_product.apply(self.source_product.source.from_numpy(v)).to_numpy()
+
+                def mvinv(v):
+                    return self.source_product.apply_inverse(self.source_product.range.from_numpy(v)).to_numpy()
+                L = LinearOperator((self.source_product.source.dim, self.source_product.range.dim), matvec=mv)
+                Linv = LinearOperator((self.source_product.range.dim, self.source_product.source.dim), matvec=mvinv)
+                return eigsh(L, sigma=0, which="LM", return_eigenvectors=False, k=1, OPinv=Linv)[0]
```

Comment:
While we are refactoring the code: I dislike that we need to rely on scipy here and to numpy conversion. Unfortunately, pyMOR has only `eigs` and not `eigsh` implemented (I believe @lbalicki wanted to change that :wink:), but we could actually use a `random_ghep` for that, couldn't we?

Reply:
I actually like the idea of using `random_ghep`. I will add this. Talking about which, I am a bit unsure about its implementation. It seems to me that some variables are allocated but never actually used. I will change that as well

Reply:
On second thoughts, we actually need the smallest eigenvalue of `S` which is not currently implemented with `random_ghep`. I am also not sure if randomized methods are the right method for that task in general..

Reply:
I have just refactored the GHEP function and I think there even was a mistake in the two-pass version.

Reply:
> On second thoughts, we actually need the smallest eigenvalue of `S` which is not currently implemented with `random_ghep`. I am also not sure if randomized methods are the right method for that task in general..

Yes, `random_ghep` is lacking a shift-invert mode. But one could simply pass `InverseOperator(S)`.